safenode/src/tests_e2e/data_with_churn.rs

```diff
@@ -135,14 +137,30 @@ async fn data_availability_during_churn() -> Result<()> {
     }
 
     println!();
-    println!("Test stopped after running for {:?}.", start_time.elapsed());
+    println!(
+        "Test stopping after running for {:?}.",
+        start_time.elapsed()
+    );
     println!();
 
     let failed = failures.read().await;
     if failed.len() > 0 {
         bail!("{} failure/s in test: {:?}", failed.len(), failed.values());
     }
 
+    // The churning of storing_chunk/querying_chunk are all random,
+    // which will have a high chance that newly stored chunk got queried BEFORE
+    // the original holders churned out.
+    // i.e. the test may pass even without any replication
+    // Hence, we carry out a final round of query all data to confirm storage.
+    for net_addr in content.read().await.iter() {
+        println!("Final querying content at {net_addr:?}");
+        assert!(
```

Comment:
I'm not sure that last check we are adding to the churn test is valid...the test retries (set to 5 times now) when querying any piece of content failed, this is because a content is expected to fail at some point in time if one of the nodes holding a replica has been churned, so that's why the test retries trying to make sure a piece of content is effectively gone.
Now this last check we are adding is querying all content created but we don't retry, I think it's expected to fail since, even if replication is working among nodes, it could have been that 1 sec. earlier one of the content's replica was churned and the replication didn't take place just yet, but that doesn't mean it was not gonna take place a few secs later.

Perhaps for this final check we want to also retry each failure (`MAX_NUM_OF_QUERY_ATTEMPTS` times), with proper delays in between (`CONTENT_QUERY_RATIO_TO_CHURN`), till we can say it's lost...?....

Plus, maybe we want to stop the churning and content creation tasks when doing this last check.

Reply:
This round of final query check is AFTER those data got queried and experienced 5 times query already.

Reply:
> This round of final query check is AFTER those data got queried and experienced 5 times query already.

no, the queries are done on random content, the querying task doesn't query all content but pick one at random each time

> Actually, I think adding a retry may cover a potential issue ?

no, we may be giving false negatives since replication may be taking place still, just that it needs more time to be triggered

> The churning that happens in-between or simutaneous shall not affect so much (fails all 8 copies)

this is adding randomness to the result, you need to stop the churning as it could be they churn those nodes holding the content you are querying, just to make sure and remove the randomness where you don't want it
..in fact, keeping the churning task may affect replication in some fashion you don't want to affect at that last check...

Reply:
This should be sorted by my commits updating the churn test 👍